app/src/main/java/com/github/se/cyrcle/model/user/UserViewModel.kt

```diff
@@ -95,20 +111,35 @@ class UserViewModel(
   }
 
   /**
-   * Updates a user in the Firestore database.
+   * Updates a user in the Firestore database. If a user has Uri for the profile picture, it will
+   * upload the image to the cloud storage and then delete the Uri.
    *
    * @param user the user to update
    */
-  fun updateUser(user: User) {
-    userRepository.updateUser(
-        user,
-        onSuccess = { Log.d("UserViewModel", "User updated") },
-        onFailure = { exception ->
-          Log.e(
-              "com.github.se.cyrcle.model.user.UserViewModel",
-              "Failed to update user: ${user.public.userId}",
-              exception)
-        })
+  fun updateUser(user: User, context: Context? = null) {
+    if (user.public.userId != currentUser.value?.public?.userId) {
+      Log.e("UserViewModel", "Trying to update a user that is not the current user")
+      return
+    }
+
+    Log.d("UserViewModel", "Updating user: $user")
+    // First update the user object in the view model.
+    setCurrentUser(user)
+    // Then update the user object in the database
+    if (user.localSession?.profilePictureUri == null || context == null) {
+      userRepository.updateUser(
+          user,
+          onSuccess = { Log.d("UserViewModel", "User updated in the database") },
+          onFailure = { exception ->
+            Log.e(
+                "com.github.se.cyrcle.model.user.UserViewModel",
+                "Failed to update user: ${user.public.userId}",
+                exception)
+          })
+    } else {
+      Log.d("UserViewModel", "User has a profile picture to upload")
+      uploadProfilePicture(context)
```

Comment:
What happens if the user modify his picture but at the same time change his username ? Shouldn't you update the user anyway ?

Reply:
Ah, I made this confusing sorry, the `uploadProfilePicture` body calls `updateUser` once the profile picture has been uploaded. 

Do you think adding a comment on line 140 would make it less confusing ?

Reply:
I think so (I mean, in general if there's a risk it may lead to confusion, better safe than sorry w/ comments :)) )

Reply:
The comment would be useful for us since we are going deeper in the code if it is okay for you could be a good idea.

Reply:
I added some explanation in the latest commit